grid/src/Cabana_Grid_Splines.hpp

```diff
@@ -451,6 +451,586 @@ struct Spline<3>
+      \param x0 The coordinate at which to evaluate the spline stencil.
+      \param indices The indices of the stencil.
+    */
+    template <class Scalar>
+    KOKKOS_INLINE_FUNCTION static void stencil( const Scalar x0,
+                                                int indices[num_knot] )
+    {
+        indices[0] = static_cast<int>( x0 ) - 2;
+        indices[1] = indices[0] + 1;
+        indices[2] = indices[1] + 1;
+        indices[3] = indices[2] + 1;
+        indices[4] = indices[3] + 1;
+    }
+
+    /*!
+      \brief Calculate the value of the spline at all knots.
+      \param x0 The coordinate at which to evaluate the spline in the logical
+      grid space.
+      \param values Basis values at the knots. Ordered from lowest to highest
+      in terms of knot location.
+    */
+    template <typename Scalar>
+    KOKKOS_INLINE_FUNCTION static void value( const Scalar x0,
+                                              Scalar values[num_knot] )
+    {
+        // Constants
+        Scalar denom_2 = 1.0 / 384.0;
+        Scalar denom_1 = denom_2 * 4.0;
+        Scalar denom_0 = denom_2 * 2.0;
+
+        // Knot at i - 2
+        Scalar xn = x0 - static_cast<int>( x0 );
+        xn -= 0.5;
+        values[0] =
+            ( 1.0 +
+              xn * ( -8.0 + xn * ( 24.0 + xn * ( -32.0 + xn * ( 16.0 ) ) ) ) ) *
+            denom_2;
+
+        // Knot at i - 1
+        values[1] =
+            ( 19.0 + xn * ( -44.0 +
+                            xn * ( 24.0 + xn * ( 16.0 + xn * ( -16.0 ) ) ) ) ) *
+            denom_1;
+
+        // Knot at i
+        values[2] =
+            ( 115.0 + xn * xn * ( -120.0 + xn * xn * ( 48.0 ) ) ) * denom_0;
+
+        // Knot at i + 1
+        values[3] =
+            ( 19.0 + xn * ( 44.0 + xn * ( 24.0 + xn * ( -16.0 +
+                                                        xn * ( -16.0 ) ) ) ) ) *
+            denom_1;
+
+        // Knot at i + 2
+        values[4] =
+            ( 1.0 +
+              xn * ( 8.0 + xn * ( 24.0 + xn * ( 32.0 + xn * ( 16.0 ) ) ) ) ) *
+            denom_2;
+    }
+
+    /*!
+      \brief Calculate the value of the gradient of the spline in the
+      physical frame.
+      \param x0 The coordinate at which to evaluate the spline in the logical
+      grid space.
+      \param rdx The inverse of the physical distance between grid locations.
+      \param gradients Basis gradient values at the knots in the physical frame.
+      Ordered from lowest to highest in terms of knot location.
+    */
+    template <typename Scalar>
+    KOKKOS_INLINE_FUNCTION static void
+    gradient( const Scalar x0, const Scalar rdx, Scalar gradients[num_knot] )
+    {
+        // Constants
+        Scalar denom_2 = 1.0 / 48.0;
+        Scalar denom_1 = denom_2 * 2.0;
+        Scalar denom_0 = denom_2 * 12.0;
+
+        // Knot at i - 2
+        Scalar xn = x0 - static_cast<int>( x0 ); // - 1.5;
+        xn -= 0.5;
```

Comment:
Same as above, pre-calculate the powers of `xn` to prevent recomputing the values for each gradient and cleanup the formatting using the precomputed values